src/liballoc/sync.rs

```diff
@@ -1131,6 +1131,55 @@ impl<T: ?Sized> Weak<T> {
             Some(unsafe { self.ptr.as_ref() })
         }
     }
+
+    /// Returns true if the two `Weak`s point to the same value (not just values
+    /// that compare as equal).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(weak_ptr_eq)]
+    /// use std::sync::{Arc, Weak};
+    ///
+    /// let first_rc = Arc::new(5);
+    /// let first = Arc::downgrade(&first_rc);
+    /// let second = Arc::downgrade(&first_rc);
+    ///
+    /// assert!(Weak::ptr_eq(&first, &second));
+    ///
+    /// let third_rc = Arc::new(5);
+    /// let third = Arc::downgrade(&third_rc);
+    ///
+    /// assert!(!Weak::ptr_eq(&first, &third));
+    /// ```
+    ///
+    /// When using `Weak::new` the `Weak` is never equal to any other `Weak`
+    /// until it points to actual data (e.g. by setting it to a value created by
+    /// calling [`Arc::downgrade`]).
```

Comment:
I think this documentation is slightly confusion. Outside of `std`, I don't think there is any way to turn a reference obtained through `Weak::new` into a reference to something real. I think it should just say something like:

> References obtained through `Weak::new` are not equal to any other references, including other invocations of `Weak::new`.

(same applies to `rc::Weak`)

Reply:
> Outside of std, I don't think there is any way to turn a reference obtained through Weak::new into a reference to something real.

I'm not sure what you mean by that. You can overwrite the value, e.g.

```rust
let mut weak_value = Weak::new();
let value = Rc::new(5);
weak_value = Rc::dowgrade(&value);
```

As for the comparing two `Weak::new()` with another I agree, but isn't the example enough? Currently no pointer created via `Weak::new()` is equal to any other pointer, including pointers also created via `Weak::new()`. That is what the documentations currently says.

Reply:
The way the documentation is written implies that you can take a `Weak::new` and somehow turn it into something that's not a dangling pointer. For example some `fn(self: &mut Weak<T>)`. Assignment and `mem::replace` don't count.

Reply:
Sorry @jethrogb I still don't really see your point. You keep mentioning "dangling pointer", but the public documentation doesn't. Furthermore you can turn a variable with the current value of `Weak::new` into something else by using plain assignment. That is also possible in any function that takes `&mut Weak<T>`, so I don't get why that doesn't "count".

Reply:
"dangling pointer" and `Weak::new` are equivalent.

> Furthermore you can turn a variable with the current value of Weak::new into something else by using plain assignment.

Dropping the old value and writing something new in its place is *not* the same as changing the value.

Reply:
> @jethrogb "dangling pointer" and Weak::new are not equivalent. That is only how the current API is implementated. Previous versions of Rust (not sure what version changed it) actually allocated on the heap when calling Weak::new. A change was made to not allocate by default for most (collection) types, including Weak. However this happend without a change in the API. This means we could go back to that implementation and then Weak::new and "dangling pointer" aren't equivalent anymore.

Ok fine, I was using them equivalently, but I meant `Weak::new`. This point is completely orthogonal to my main point about the phrasing of the documentation.

Reply:
> When using `Weak::new` the `Weak` is never equal to any other `Weak` until it points to actual data

The "until it points to actual data" part clearly implies you can take a `Weak::new` and somehow change it to a value that is real. Dropping the old value and moving something new in its place is *not* the same as changing the value.